crates/pyndakaas/python/pindakaas/solver.py

```diff
@@ -0,0 +1,110 @@
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+from typing import ContextManager, Iterable, Iterator, Optional
+from datetime import timedelta
+
+from .encoding import ClauseDatabase, Constraint
+from .pindakaas import Lit, Encoder
+from .pindakaas.solver import CaDiCaLInner, KissatInner, Status
+
+
+class Result(ABC):
+    @property
+    @abstractmethod
+    def status(self) -> Status: ...
+
+    @abstractmethod
+    def value(self, var: Lit) -> Optional[bool]: ...
+
+    @abstractmethod
+    def failed(self, var: Lit) -> Optional[bool]: ...
+
+
+class Solver(ClauseDatabase):
+    def _set_time_limit(self, limit: Optional[timedelta]):
+        if limit is not None:
+            raise NotImplementedError("Solver does not support setting a time limit")
+
+    @abstractmethod
+    def _solve_assuming(self, assumptions: Iterable[Lit]) -> ContextManager[Result]: ...
+
+    @contextmanager
+    def solve(
+        self, assumptions: Iterable[Lit] = [], time_limit: Optional[timedelta] = None
```

Comment:
using `= []` might be a bit dangerous with python's default argument behavior.

Reply:
I don't think it is a problem because we never actually mutate it, but if you prefer to use the `None` approach, we can change this.